ansible/roles/test/files/ptftests/advanced-reboot.py

```diff
@@ -947,31 +976,131 @@ def check_alive(self):
 
-        is_alive      = nr_from_s > self.nr_pc_pkts * 0.7 and nr_from_l > self.nr_vl_pkts * 0.7
-        is_asic_weird = nr_from_s > self.nr_pc_pkts        or nr_from_l > self.nr_vl_pkts
-        # we receive more, then sent. not populated FDB table
+    def get_asic_state(self):
+        with self.asic_state_lock:
+            state = self.asic_state
+        return state
+
+
+    def get_asic_state_time(self, state):
+        with self.asic_state_lock:
+            time = self.asic_state_time[state]
+        return time
+
+
+    def set_asic_state(self, state, reachability=0):
+        with self.asic_state_lock:
+            self.asic_state             = state
+            self.asic_state_time[state] = datetime.datetime.now()
+
+
+    def get_asic_vlan_reachability(self):
+        return self.asic_vlan_reach
+
+
+    def asic_start_recording_vlan_reachability(self):
+        with self.asic_state_lock:
+            self.asic_vlan_reach = []
+            self.recording       = True
+
+
+    def asic_stop_recording_vlan_reachability(self):
+        with self.asic_state_lock:
+            self.recording = False
+
+
+    def try_record_asic_vlan_recachability(self, t1_to_vlan):
+        with self.asic_state_lock:
+            if self.recording:
+                self.asic_vlan_reach.append(t1_to_vlan)
+
+    def get_cpu_state(self):
+        with self.cpu_state_lock:
+            state = self.cpu_state
+        return state
+
+
+    def get_cpu_state_time(self, state):
+        with self.cpu_state_lock:
+            time = self.cpu_state_time[state]
+        return time
+
+
+    def set_cpu_state(self, state):
+        with self.cpu_state_lock:
+            self.cpu_state             = state
+            self.cpu_state_time[state] = datetime.datetime.now()
+
+
+    def log_asic_state_change(self, reachable, partial=False, t1_to_vlan=0):
+        old = self.get_asic_state()
+
+        if reachable:
+            state = 'up' if not partial else 'partial'
+        else:
+            state = 'down'
+
+        self.try_record_asic_vlan_recachability(t1_to_vlan)
+
+        if old != state:
+            self.log("ASIC state transition from %s to %s (%d)" % (old, state, t1_to_vlan))
+            self.set_asic_state(state)
+
+
+    def log_cpu_state_change(self, reachable, partial=False):
+        old = self.get_cpu_state()
+
+        if reachable:
+            state = 'up' if not partial else 'partial'
+        else:
+            state = 'down'
+
+        if old != state:
+            self.log("CPU state transition from %s to %s" % (old, state))
+            self.set_cpu_state(state)
+
+
+    def reachability_watcher(self):
+        # This function watches the reachability of the CPU port, and ASIC. It logs the state
+        # changes for future analysis
+
+        while self.watching:
+            vlan_to_t1, t1_to_vlan = self.ping_data_plane(True)
+            reachable              = (t1_to_vlan  > 0 and vlan_to_t1 > 0  and
```

Comment:
I added the new condition to be more robust. In case we change total packet count in the future, and 0.7 of total packet drops to 0, the check will prevent false negative.